File: web/src/api/message.ts
export enum MessageType {
  RequestGame = "req_game",
  WaitTimeout       = "wait_timeout",
  SessionCreated    = "session_created",
  WorkerNotFound    = "worker_not_found",
  RequestGameFailed = "request_game_failed",
  SDP          = "sdp",
  ICECandidate = "ice_candidate"
}
/* 
const MessageRequestGame = "req_game"
const MessageWaitTimeout       = "wait_timeout"
const	MessageSessionCreated    = "session_created"
const	MessageWorkerNotFound    = "worker_not_found"
const	MessageRequestGameFailed = "request_game_failed"

const MessageSDP          = "sdp"
const	MessageICECandidate = "ice_candidate" */

export type WsMessage = {
  session_id: string;
  sender_id: string;
  receiver_ids: string[];
  type: string;
  payload: string;
}
export type RequestGamePayload = {
  app_name :string;
  mode :string;
}

export type SessionCreatedPayload = {
  worker_id :string;
  session_id:string;
  player_ids: string[];
}

File: coordinator/message.go
package coordinator

import (
	"encoding/json"

	"github.com/khoakmp/brgame/api"
)

type MessageBuilder struct {
}

var MsgBuilder MessageBuilder

func (mb *MessageBuilder) SessionCreated(workerID, sessionID string, playerIDs []string) *api.Message {
	payload := api.SessionCreatedPayload{
		WorkerID:  workerID,
		SessionID: sessionID,
		PlayerIDs: playerIDs,
	}
	buf, _ := json.Marshal(payload)

	return &api.Message{
		SessionID:   sessionID,
		SenderID:    "coordinator_",
		ReceiverIDs: nil,
		Type:        api.MessageSessionCreated,
		Payload:     string(buf),
	}
}

func (mb *MessageBuilder) WorkerNotFound() *api.Message {
	return &api.Message{
		Type:     api.MessageWorkerNotFound,
		SenderID: "coordinator_",
		Payload:  "serivce unavailable",
	}
}

func (mb *MessageBuilder) StartSession(sessionID, appName string, clientIDs []string, workerID string) *api.Message {
	payload := api.StartSessionPayload{
		ClientIDs: clientIDs,
		AppName:   appName,
	}

	buf, _ := json.Marshal(payload)
	startMessage := api.Message{
		SessionID:   sessionID,
		SenderID:    "coordinator_",
		ReceiverIDs: nil,
		Type:        api.MessageStartSession,
		Payload:     string(buf),
	}
	return &startMessage
}
